apps/aecore/lib/aecore/peers/worker.ex

```diff
@@ -113,44 +144,57 @@ defmodule Aecore.Peers.Worker do
   is updated if the one in the latest GET /info request is different.
   """
   def handle_call(:check_peers, _from, %{peers: peers} = state) do
-    filtered_peers = :maps.filter(fn(peer, _) ->
-        case Client.get_info(peer) do
-          {:ok, info} -> info.genesis_block_hash == genesis_block_header_hash()
-          _ -> false
+    filtered_peers = :maps.filter(fn(_, %{uri: uri}) ->
+        case Client.get_info(uri) do
+          {:ok, info} ->
+            info.genesis_block_hash == genesis_block_header_hash()
+          _ ->
+            false
         end
       end, peers)
-
     updated_peers =
-      for {peer, current_block_hash} <- filtered_peers, into: %{} do
-        {_, info} = Client.get_info(peer)
-        if info.current_block_hash != current_block_hash do
-          {peer, info.current_block_hash}
+      for {nonce, %{uri: uri, latest_block: latest_block}} <- filtered_peers, into: %{} do
+        {_, info} = Client.get_info(uri)
+        if(info.current_block_hash != latest_block) do
+          {nonce, %{uri: uri, latest_block: info.current_block_hash}}
         else
-          {peer, current_block_hash}
+          {nonce, %{uri: uri, latest_block: latest_block}}
         end
       end
 
     removed_peers_count = Enum.count(peers) - Enum.count(filtered_peers)
     if removed_peers_count > 0 do
-      Logger.info(fn -> "#{Enum.count(peers) - Enum.count(filtered_peers)} peers were removed after the check" end)
+      Logger.info(fn -> "#{removed_peers_count} peers were removed after the check" end)
     end
 
     {:reply, :ok, %{state | peers: updated_peers}}
   end
 
   def handle_call(:all_peers, _from, %{peers: peers} = state) do
-    {:reply, peers, %{state | peers: peers}}
+    {:reply, peers, state}
   end
 
   ## Async operations
 
   def handle_cast({:broadcast_to_all, {type, data}}, %{peers: peers} = state) do
-    send_to_peers(type, data, Map.keys(peers))
+    peer_uris = peers
+      |> Map.values()
+      |> Enum.map(fn(%{uri: uri}) -> uri end)
+    send_to_peers(type, data, peer_uris)
     {:noreply, state}
   end
 
-  def handle_cast({:schedule_add_peer, uri}, state) do
-    {:reply, _, state} = add_peer(uri, state)
+  def handle_cast({:schedule_add_peer, uri, nonce}, %{peers: peers} = state) do
+    already_have_peer = Map.has_key?(peers, nonce)
+    state =
+      case already_have_peer do
```

Comment:
In general, I would not use `case ... do` statements to match on booleans. An if/else statement works just fine. 

```elixir
if Map.has_key?(peers, nonce) do
  {:noreply, state}
else
  {:reply, _, newstate} = add_peer(uri, state)
  {:noreply, newstate}
end
```